### PR TITLE
Terminal prints

### PR DESCRIPTION
**Pull request by jvollhueter**
Here is the pull request we have talked about in #152 .

1. The first commit the terminal print of time steps is transformed in years and formatted to two decimals.
2. In the second commit the terminal print gets optionally and you can choose whether you want days or years as unit. Because of Backward compatibility of old xml project files the new introduced section "terminal_prints" does not have to be included.
3. The last commit gives an overview how many time steps already got calculated.

I think the way how I implemented the passing of the variable that gets read in the xml file "terminal_prints" is not done in a good way. Do you have suggestions to improve this?

### CODE REVIEW

File: ProjectLib/XMLtoProject.py [hunk 1]
```diff
@@ -27,6 +27,7 @@ def __init__(self, **args):
         self.addTreeTimeLoop()
         self.addVisualization()
         self.addTreeOutput()
+        self.addTerminalPrints()
```
Comment (jbathmann) [inline ProjectLib/XMLtoProject.py:30]: see other comments concerning this specific addition

File: TimeLoopLib/TreeDynamicTimeLoop.py [hunk 2]
```diff
@@ -26,9 +26,22 @@ def getNextTimeStepBoundaries(self):
...
+                      '\nCalculated timesteps: ' +
+                      str(int((self.loop.t_2/self.loop.t_end)*100)) + ' %')
+            elif self.terminal_print == 'days':
+                print("Next time step to propagate" +
+                      " tree population with starting time " +
+                      '%4.2f' % (float(self.loop.t_1)/86400) +
+                      " d and end time " +
+                      '%4.2f' % (float(self.loop.t_2)/86400) + " d." +
+                      '\nCalculated timesteps: ' +
+                      str(int((self.loop.t_1/self.loop.t_end)*100)) + ' %')
```
Comment (jbathmann) [inline TimeLoopLib/TreeDynamicTimeLoop.py:44]: What is the default? Can I turn it off?

File: TimeLoopLib/TreeDynamicTimeLoop.py [hunk 3]
```diff
@@ -26,9 +26,22 @@ def getNextTimeStepBoundaries(self):
...
         while (self.loop.step_on):
-            print("Next time step to propagate" +
-                  " tree population with starting time " + str(self.loop.t_1) +
-                  " and end time " + str(self.loop.t_2) + ".")
+            if self.terminal_print == 'years':
+                print("Next time step to propagate" +
+                      " tree population with starting time " +
+                      '%4.2f' % (float(self.loop.t_1)/86400/365.25) +
+                      " a and end time " +
+                      '%4.2f' % (float(self.loop.t_2)/86400/365.25) + " a." +
```
Comment (jbathmann) [inline TimeLoopLib/TreeDynamicTimeLoop.py:34]: I think I always used 365 days for one year as it makes things easier. 

However, this makes me think of something like:

`'%4.2f' % (float(self.loop.t_1)/terminal_prompt_unit)`

wdyt?

File: test/SmallTests/Test_Setups_small/AllSimple.xml [hunk 4]
```diff
@@ -54,5 +54,7 @@
     <tree_output>
         <type> NONE </type>
     </tree_output>
+    <terminal_prints>
+        days
+    </terminal_prints>
```
Comment (jbathmann) [inline test/SmallTests/Test_Setups_small/AllSimple.xml:59]: see above

File: ProjectLib/XMLtoProject.py [hunk 5]
```diff
@@ -68,7 +67,7 @@ def addTreeOutput(self):
     def findChild(self, parent, key):
         child = parent.find(key)
         if child is None:
-            raise KeyError("key '" + key + "' is missing in project file")
+            raise KeyError("key " + key + " is missing in project file")
```
Comment (jbathmann) [inline ProjectLib/XMLtoProject.py:70]: why do you want to remove the " ' "?

File: TimeLoopLib/TreeDynamicTimeLoop.py [hunk 6]
```diff
@@ -13,6 +13,10 @@ def __init__(self, args):
             self.iniSimpleTimeStepping(args)
         else:
             raise KeyError("Required time stepping not implemented.")
+        try:
+            self.terminal_print = args.find('terminal_print').text
```
Comment (jbathmann) [inline TimeLoopLib/TreeDynamicTimeLoop.py:17]: Doesn't this always work and just return a "None" if 'terminal_print' is not further specified?

**Comment by jbathmann** [comment]
@jvollhueter could you please rebase?

File: TimeLoopLib/TreeDynamicTimeLoop.py (hunk 2 above)
Comment (jbathmann) [reply TimeLoopLib/TreeDynamicTimeLoop.py:51]: :+1: 

Can you maybe add this to some inline comments?

**Comment by jbathmann** [comment]
Looks good!
Can you please add prj file documentation of the new tags? 

- Add new tags to Documentation/ProjectFile
- run generate_tag_documentation in ./Documentation

**Comment by jbathmann** [comment]
@mcwimm can we merge? :)

File: TimeLoopLib/TreeDynamicTimeLoop.py (hunk 6 above)
Comment (mcwimm) [reply TimeLoopLib/TreeDynamicTimeLoop.py:17]: I don't get the prupose of those prints

**Review by mcwimm — COMMENTED** [review]
I like the new option. However, some documentation would be helpful (i.e. tag and tag options).

File: TimeLoopLib/TreeDynamicTimeLoop.py (hunk 3 above)
Comment (mcwimm) [reply TimeLoopLib/TreeDynamicTimeLoop.py:41]: imo, a good idea

File: TimeLoopLib/TreeDynamicTimeLoop.py (hunk 2 above)
Comment (jvollhueter) [reply TimeLoopLib/TreeDynamicTimeLoop.py:51]: Done

File: TimeLoopLib/TreeDynamicTimeLoop.py (hunk 3 above)
Comment (jvollhueter) [reply TimeLoopLib/TreeDynamicTimeLoop.py:41]: Done

**Comment by jvollhueter** [comment]
I tried to add the new tag in the documentation. Is there something like a preview for the documentation? I'm not sure whether I did correctly...

**Comment by mcwimm** [comment]
> I tried to add the new tag in the documentation. Is there something like a preview for the documentation? I'm not sure whether I did correctly...

You can run `doxygen Doxyfile` from the root directory. This builds the documentation locally. It might take a few minutes.

**Comment by jbathmann** [comment]
Thanks for the documentation. Looks good.

**Comment by jvollhueter** [comment]
There was an error in the documentation, but now it looks good. I think it's ready now!

**Pull request merged**